prow/tide/gerrit.go

```diff
@@ -0,0 +1,312 @@
+	return p.gc.GetBranchRevision(org, repo, ref)
+}
+
+// headContexts gets the status contexts for the commit with OID ==
+// pr.HeadRefOID
+//
+// Assuming all submission requirements are already met as the PRs queried are
+// already submittable. So the focus here is to ensure that all prowjobs were
+// tested against latest baseSHA.
+// Prow parses baseSHA from the `Description` field of a context, will make sure
+// that all Prow jobs that vote to required labels are represented here.
+func (p *GerritProvider) headContexts(crc *CodeReviewCommon) ([]Context, error) {
+	var res []Context
+
+	selector := map[string]string{
+		kube.GerritRevision:   crc.HeadRefOID,
+		kube.ProwJobTypeLabel: string(prowapi.PresubmitJob),
+		kube.OrgLabel:         crc.Org,
+		kube.RepoLabel:        crc.Repo,
+		kube.PullLabel:        strconv.Itoa(crc.Number),
+	}
+	var pjs v1.ProwJobList
+	if err := p.pjclientset.List(context.Background(), &pjs, ctrlruntimeclient.MatchingLabels(selector)); err != nil {
+		return nil, fmt.Errorf("Cannot list prowjob with selector %v", selector)
+	}
+
+	// keep track of latest prowjobs only
+	latestPjs := make(map[string]*prowapi.ProwJob)
+	for _, pj := range pjs.Items {
+		pj := pj
+		exist, ok := latestPjs[pj.Spec.Context]
+		if ok && exist.CreationTimestamp.After(pj.CreationTimestamp.Time) {
+			continue
+		}
+		latestPjs[pj.Spec.Context] = &pj
+	}
+
+	for _, pj := range latestPjs {
+		res = append(res, Context{
+			Context:     githubql.String(pj.Spec.Context),
+			Description: githubql.String(config.ContextDescriptionWithBaseSha(pj.Status.Description, pj.Spec.Refs.BaseSHA)),
+			State:       githubql.StatusState(pj.Status.State),
+		})
+	}
+
+	return res, nil
+}
+
+func (p *GerritProvider) mergePRs(sp subpool, prs []CodeReviewCommon, dontUpdateStatus *threadSafePRSet) error {
+	p.logger.Info("The merge function hasn't been implemented yet, just logging for now.")
+	return nil
+}
+
+// GetTideContextPolicy gets context policy defined by users + requirements from
+// prow jobs.
+func (p *GerritProvider) GetTideContextPolicy(gitClient git.ClientFactory, org, repo, branch string, baseSHAGetter config.RefGetter, crc *CodeReviewCommon) (contextChecker, error) {
+	pr := crc.Gerrit
+	if pr == nil {
+		return nil, errors.New("error not supposed to happen, crc.Gerrit cannot be nil for GerritProvider")
+	}
+
+	required := sets.NewString()
+	requiredIfPresent := sets.NewString()
+	optional := sets.NewString()
+
+	headSHAGetter := func() (string, error) {
+		return crc.HeadRefOID, nil
+	}
+	presubmits, err := p.cfg().GetPresubmits(gitClient, org+"/"+repo, baseSHAGetter, headSHAGetter)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get presubmits: %w", err)
+	}
+
+	requireLabels := sets.NewString()
+	for l, info := range pr.Labels {
+		if !info.Optional {
+			requireLabels.Insert(l)
+		}
+	}
+
+	// automatically generate required and optional entries for Prow Jobs
```

Comment:
Nit: Remove `automatically` here? If automation is important somehow, maybe explain how this is different from some other known way of generating these required and optional entries. I.e., is there a manual (more involved) way of doing the same thing?

Reply:
done